Support for Zendesk

mozts2005 opened the pull request:
I have added Support for Zendesk. I am happy to help in testing validation as I have an account with Zendesk that is used for Open-source project testing.

kevinchalet commented on src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml:
```diff
@@ -2153,6 +2153,27 @@
+                     TokenEndpoint="https://{settings.Tenant}.zendesk.com/oauth/tokens"
+                     UserInfoEndpoint="https://{settings.Tenant}.zendesk.com/api/v2/users/me">
+        <CodeChallengeMethod Value="S256" />
+        <GrantType Value="authorization_code" />
```
```suggestion
```

When `authorization_code` is the only grant type supported, adding it explicitly isn't required as it's always added by the provider generator when you don't explicitly set a list of grant types.

kevinchalet commented on src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml:
```diff
@@ -2153,6 +2153,27 @@
+    </Environment>
+    <Setting PropertyName="Tenant" ParameterName="tenant" Type="String" Required="true"
+             Description="Gets or sets the tenant used to identify the Zendesk instance." />
+    <Setting PropertyName="ClientType" ParameterName="kind" Type="String" Required="false"
+         DefaultValue="confidential" Description="Client Type: public or confidential. Public OAuth clients are applications that run in environments where credentials cannot be securely stored, such as mobile and web apps. These clients are required to use PKCE. Confidential OAuth clients run on secure servers where their credentials can be kept secure. These clients can use PKCE, client secret, ...
```
```suggestion
```

This setting shouldn't be needed (and since you didn't write custom code to do something based on its value, it really has no effect whatsoever) 😃

mozts2005 replied:
Should the value not be automatically added to the code request as it has both a property name and a parameter name? if not where should I update to include this value as this is required or the client secret is required. I would like to also add validation for configuration.

kevinchalet replied:
> Should the value not be automatically added to the code request as it has both a property name and a parameter name?

Nope (note: the "parameter name" here refers to the name of the parameter used in the C# builder method, not an OIDC request parameter).

> if not where should I update to include this value as this is required or the client secret is required. I would like to also add validation for configuration.

You don't need that parameter: when no client secret or signing credentials is attached to a client registration, OpenIddict detects it when negotiating the client authentication method and sends the token request without a secret attached.

There's no need for special validation logic for Zendesk, it behaves like any other standard provider in this regard.

kevinchalet commented on src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml:
```diff
@@ -2153,6 +2153,27 @@
+      </Configuration>
+    </Environment>
+    <Setting PropertyName="Tenant" ParameterName="tenant" Type="String" Required="true"
+             Description="Gets or sets the tenant used to identify the Zendesk instance." />
```
```suggestion
             Description="Gets or sets the tenant used to identify the Zendesk instance" />
```

(the . is automagically added by the generator)

kevinchalet commented on src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml:
```diff
@@ -2153,6 +2153,27 @@
     <Environment Issuer="https://api.login.yahoo.com/" />
   </Provider>
 
+  <Provider Name="Zendesk" Id="5577b117-1f3d-464f-b987-aabf7c0ab3f7" Documentation="https://developer.zendesk.com/api-reference/introduction/security-and-auth/#oauth-access-token">
```
```suggestion
  <Provider Name="Zendesk" Id="5577b117-1f3d-464f-b987-aabf7c0ab3f7" Documentation="https://support.zendesk.com/hc/en-us/articles/4408845965210-Using-OAuth-authentication-with-your-application">
```

Let's use https://support.zendesk.com/hc/en-us/articles/4408845965210-Using-OAuth-authentication-with-your-application instead of https://developer.zendesk.com/api-reference/introduction/security-and-auth/#oauth-access-token as the last one doesn't contain anything really useful.

kevinchalet reviewed (commented):
Thanks for your PR! ❤️ 

> I am happy to help in testing validation as I have an account with Zendesk that is used for Open-source project testing.

Yes please! (I don't have a Zendesk account and won't be able to test it myself)

kevinchalet commented:
Hey @mozts2005,

If you have any question or need help to complete this PR, please let me know 😃 

Cheers.

mozts2005 committed:
Support for Zendesk

mozts2005 commented:
I made the changes you requested. 
I also added support for revocation of the token. 
Please let me know if you would like to handle the revocation authentication differently. I did it this way as the same type of override is done for the Trovo userinfo endpoint authentication. https://github.com/openiddict/openiddict-core/blob/ff752efe09b319ee930dd181a77edd34a69a0f48/src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationHandlers.Userinfo.cs#L200
But I was thinking that if you wanted this could be added as a new EndpointAuthMethod if you prefer. 


Next, I added some mappings for to the UserInfo based on Zendesk endpoint results but if you would like I can remove them or add more.

kevinchalet commented:
> Please let me know if you would like to handle the revocation authentication differently.

No that's fine 👍🏻

kevinchalet commented on src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationHandlers.cs:
```diff
@@ -1362,6 +1362,9 @@ public ValueTask HandleAsync(ProcessAuthenticationContext context)
                 ProviderTypes.Shopify => (string?) context.TokenResponse?["associated_user"]?["email"],
+                
+                // Zendesk returns the emails address as a custom "user/email" node:
+                ProviderTypes.Zendesk => (string?) context.UserInfoResponse?["user"]?["email"],
```
Looks like the userinfo details are wrapped in a `user` node? If so, we'll need to unwrap the userinfo response so all the claims can be correctly mapped: https://documentation.openiddict.com/guides/contributing-a-new-web-provider#unwrap-userinfo-responses-if-necessary

Once done, you'll need to remove the `?["user"]` part of these new lines.

mozts2005 replied:
updated

kevinchalet commented:
Thanks. Could you please attach a screenshot of the returned claims (visible in the output generated by the console client sandbox)? Feel free to blur out personal/sensitive details.

Were you able to test everything? Revocation works fine?

mozts2005 commented:
> Thanks. Could you please attach a screenshot of the returned claims (visible in the output generated by the console client sandbox)? Feel free to blur out personal/sensitive details.
> 
> Were you able to test everything? Revocation works fine?

yes, revocation works and here is the screenshot.
<img width="1705" alt="ZendeskAfterUnwrap" src="https://github.com/user-attachments/assets/34fabf91-0c2a-40e4-8285-dfc7377b736d">

kevinchalet committed:
Remove the token to revoke from the request payload after attaching it to the Authorization header

kevinchalet committed:
Add the ASCII art header

kevinchalet commented:
I pushed a commit that removes the `token` parameter from the revocation request (since Zendesk doesn't expect it and requires using bearer authentication instead). Could you please give it a try to ensure it still works fine with this change?

kevinchalet committed:
Remove an unnecessary extra line

mozts2005 commented:
> I pushed a commit that removes the `token` parameter from the revocation request (since Zendesk doesn't expect it and requires using bearer authentication instead). Could you please give it a try to ensure it still works fine with this change?


all changes made look good to me after testing.
![image](https://github.com/user-attachments/assets/180289d9-28d8-4992-84a3-6c54d850d7d5)

kevinchalet commented on src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationHandlers.Userinfo.cs:
```diff
@@ -406,8 +406,9 @@ public ValueTask HandleAsync(ExtractUserInfoResponseContext context)
 
-                    // Fitbit and Todoist return a nested "user" object.
-                    ProviderTypes.Fitbit or ProviderTypes.Todoist => new(context.Response["user"]?.GetNamedParameters() ??
+                    // Fitbit, Todoist, and Zendesk return a nested "user" object.
```
```suggestion
                    // Fitbit, Todoist and Zendesk return a nested "user" object.
```

kevinchalet committed:
Update src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationHandlers.Userinfo.cs

kevinchalet commented:
Merged. Thanks a lot for your contribution, @mozts2005! 👏🏻
